Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -266,114 +239,187 @@ final class ConnectionPool {
-        ///     - parentPool: The pool this provider belongs to
-        init(key: ConnectionPool.Key, eventLoop: EventLoop, configuration: HTTPClient.Configuration, parentPool: ConnectionPool) {
+        ///     - pool: The pool this provider belongs to
+        init(key: ConnectionPool.Key, eventLoop: EventLoop, configuration: HTTPClient.Configuration, pool: ConnectionPool) {
             self.eventLoop = eventLoop
             self.configuration = configuration
             self.key = key
-            self.parentPool = parentPool
-            self.state = State(eventLoop: eventLoop, parentPool: parentPool, key: key)
+            self.pool = pool
         }
 
         deinit {
-            assert(self.state.activity == .closed, "Non closed on deinit")
-            assert(self.state.availableConnections.isEmpty, "Available connections should be empty before deinit")
-            assert(self.state.leased == 0, "All leased connections should have been returned before deinit")
-            assert(self.state.waiters.count == 0, "Waiters on deinit: \(self.state.waiters)")
+            assert(self.waiters.isEmpty)
+            assert(self.availableConnections.isEmpty)
+            assert(self.openedConnectionsCount == 0)
         }
 
         var description: String {
-            return "HTTP1ConnectionProvider { key: \(self.key), state: \(self.state) }"
+            return "HTTP1ConnectionProvider { key: \(self.key) }"
         }
 
-        func getConnection(preference: HTTPClient.EventLoopPreference) -> EventLoopFuture<Connection> {
-            self.activityPrecondition(expected: [.opened])
-            let action = self.stateLock.withLock { self.state.connectionAction(for: preference) }
+        var isActive: Bool {
+            return self.lock.withLock {
+                self.active
+            }
+        }
+
+        func execute(_ action: Action) {
             switch action {
-            case .leaseConnection(let connection):
-                return connection.removeIdleConnectionHandlersForLease().flatMapError { _ in
-                    connection.closeFuture.flatMap { // We ensure close actions are run first
-                        let defaultEventLoop = self.stateLock.withLock {
-                            self.state.defaultEventLoop
+            case .lease(let connection, let waiter):
+                // check if we can vend this connection to caller
+                connection.cancelIdleTimeout().flatMapError { error in
+                    // if connection is already inactive, we create a new one.
+                    if error is Connection.InactiveChannelError {
+                        self.lock.withLockVoid {
+                            self.openedConnectionsCount += 1
                         }
-                        return self.makeConnection(on: preference.bestEventLoop ?? defaultEventLoop)
+                        return self.makeConnection(on: waiter.preference.bestEventLoop ?? self.eventLoop)
```

Comment:
Is this really not racy? If we shut the whole thing down, could this create a new connection after we think we've shut everything?